### PR TITLE
Implement sprinting mechanics

### PR DESCRIPTION
**Pull request by ender-titan1**
**Brief Description of What This PR Does**

This PR implements sprinting and cell strain to the game.

To do:
- [ ] Balance
- [x] Red flash when strain is high
- [x] Strain Bar (for accessibility)
- [x] Strain Bar visibility options

Maybe:
- [ ] Tutorial
- [x] Damage cell when ATP is close to zero 

**Related Issues**

https://forum.revolutionarygamesstudio.com/t/organism-stamina-mechanics/757/7

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [ ] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Review by ARecurringProblem — APPROVED** [review]
If we're settled on the basic implementation idea then I think it's good to get this into a BOTD soon so that we can get player feedback on how it feels.

File: src/general/OptionsMenu.cs [hunk 1]
```diff
@@ -501,6 +506,12 @@ public override void _Ready()
     {
         ResolveNodeReferences(true);
 
+        foreach (var mode in typeof(Settings.StrainBarVisibility).GetEnumValues().Cast<Settings.StrainBarVisibility>())
```
Comment (hhyyrylainen) [inline src/general/OptionsMenu.cs:509]: For the options menu the usual way to setup these dropdowns is to set them in the Godot editor and not load them manually like this.

File: src/engine/Settings.cs [hunk 2]
```diff
@@ -489,6 +505,11 @@ private Settings()
         Constants.CONTROLLER_DEFAULT_DEADZONE,
     });
 
+    // Gameplay and Accessibility settings
```
Comment (hhyyrylainen) [inline src/engine/Settings.cs:508]: Isn't this in the same section in the menu as the display abilities bar?

File: simulation_parameters/Constants.cs [hunk 3]
```diff
@@ -183,12 +183,38 @@ public static class Constants
     /// </remarks>
     public const float BASE_MOVEMENT_ATP_COST = 1.0f;
 
+    public const float BASE_MOVEMENT_FORCE = 900.0f;
+
+    public const float MAX_STRAIN_PER_CELL = 100.0f;
+
+    public const float PASSIVE_STRAIN_DECREASE_PER_UPDATE = 0.5f;
```
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:190]: Rather than per update rate, any change rates should be scaled to be per second.

File: src/microbe_stage/MicrobeHUD.tscn [hunk 4]
```diff
@@ -1943,10 +2102,29 @@ __meta__ = {
...
 }
 
+[node name="Sprint" parent="ScrollContainer/HotBar" instance=ExtResource( 65 )]
+visible = false
+anchor_right = 0.0
+anchor_bottom = 0.0
+margin_left = 252.0
+margin_right = 298.0
+margin_bottom = 51.0
+hint_tooltip = "TOGGLE_ENGULF_TOOLTIP"
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.tscn:2112]: Wrong tooltip text is used here.

File: src/microbe_stage/components/StrainAffected.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    [JSONDynamicTypeAllowed]
+    public struct StrainAffected
+    {
+        /// <summary>
+        ///   True when sprinting or when strain is supposed to be otherwise generated
+        /// </summary>
+        public bool IsUnderStrain;
+
+        /// <summary>
+        ///   The current amount of strain, the maximum is calculated based on the hex size
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/StrainAffected.cs:17]: Where is the maximum calculated as per this comment?

File: src/microbe_stage/systems/MicrobeMovementSystem.cs [hunk 6]
```diff
@@ -194,6 +206,16 @@ private static float CalculateRotationSpeed(in Entity entity, ref OrganelleConta
...
                 (cellProperties.MembraneRigidity * Constants.MEMBRANE_RIGIDITY_BASE_MOBILITY_MODIFIER);
 
+            if (control.Sprinting && canSprint)
+            {
+                force *= Constants.SPRINTING_FORCE_MULTIPLIER;
+                strain.IsUnderStrain = true;
+            }
+            else
+            {
+                strain.IsUnderStrain = false;
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeMovementSystem.cs:216]: Shouldn't strain be unset when no movement keys are set? In the if checking `control.MovementDirection == Vector3.Zero`.

File: src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs [hunk 7]
```diff
@@ -141,7 +141,7 @@ protected override void Update(float delta, in Entity entity)
         /// </summary>
         private void ApplyATPDamage(CompoundBag compounds, ref Health health, ref CellProperties cellProperties)
         {
-            if (compounds.GetCompoundAmount(atp) > 0)
+            if (compounds.GetCompoundAmount(atp) > 0.1f)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs:144]: Why is this needed? I think this may have massive implications on taking ATP damage when trying to move.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    using DefaultEcs.Threading;
+    using Godot;
+    using World = DefaultEcs.World;
+
+    [With(typeof(CompoundStorage))]
+    [With(typeof(OrganelleContainer))]
+    [With(typeof(MicrobeControl))]
+    [With(typeof(StrainAffected))]
+    [RunsAfter(typeof(PhysicsBodyDisablingSystem))]
+    public sealed class StrainSystem : AEntitySetSystem<float>
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:15]: Why doesn't this system rely on `IsUnderStrain`? It seems the purpose of that variable is so that the movement system sets that and then that greatly simplifies this system by not needing to act on so much microbe-specific information.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    using DefaultEcs.System;
+    using DefaultEcs.Threading;
+    using Godot;
+    using World = DefaultEcs.World;
+
+    [With(typeof(CompoundStorage))]
+    [With(typeof(OrganelleContainer))]
+    [With(typeof(MicrobeControl))]
+    [With(typeof(StrainAffected))]
+    [RunsAfter(typeof(PhysicsBodyDisablingSystem))]
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:14]: Why does this run after the body disabling system?

File: src/microbe_stage/systems/StrainSystem.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    using Components;
+    using DefaultEcs;
+    using DefaultEcs.System;
+    using DefaultEcs.Threading;
+    using Godot;
+    using World = DefaultEcs.World;
+
+    [With(typeof(CompoundStorage))]
+    [With(typeof(OrganelleContainer))]
+    [With(typeof(MicrobeControl))]
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:12]: As far as I can tell these two components are just read from and as such should have the reads from component attributes here to mark that to make multithreading more effective related to this system.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                {
+                    ReduceStrain(ref strain);
+                }
+                else
+                {
+                    strain.StrainDecreaseCooldown -= delta;
+                    ReduceStrain(ref strain, Constants.PASSIVE_STRAIN_DECREASE_PRE_COOLDOWN_DIVISOR);
+                }
+
+                if (strain.CurrentStrain < 0)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:59]: Why not put this check and clamp inside the ReduceStrain method?

File: src/microbe_stage/systems/StrainSystem.cs (hunk 9 above)
Comment (ender-titan1) [reply src/microbe_stage/systems/StrainSystem.cs:14]: The logic behind this was that since this is supposed to run before `MicrobeMovementSystem`, and that system ran after `PhysicsBodyDisablingSystem`, I moved the `RunsAfterAttribute` to `StrainSystem`, but looking at it now it might not be correct.

File: src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs (hunk 7 above)
Comment (ender-titan1) [reply src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs:144]: This was changed as ATP never reached a low enough level for it to do damage. The bar would look empty but no damage would be done. Should've probably set it to something lower though.

File: src/microbe_stage/systems/StrainSystem.cs (hunk 9 above)
Comment (hhyyrylainen) [reply src/microbe_stage/systems/StrainSystem.cs:14]: Then set this directly to run after the movement system. The threaded run generator can handle recursive dependencies so A -> B -> C runs after attributes will ensure that A will only run after C.

File: src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs (hunk 7 above)
Comment (hhyyrylainen) [reply src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs:144]: Adjusting when the ATP is taken would be a better choice. That's what's been done in the past to ensure cells can be damaged (for example making this run before whatever system applies ATP damage, but after the process system).

Comment (ender-titan1) [reply src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs]: I ended up changing this to `0.05f`, as I couldn't get this to work otherwise. I think it's sensible to set it to that, as the difference at that point is barely noticeable and I don't think it would have any major implications.

Comment (hhyyrylainen) [reply src/microbe_stage/systems/OsmoregulationAndHealingSystem.cs]: Could you add a comment before this if to explain the number? Or actually I think it'd be pretty nice to make it a constant in Constants and add a summary comment explaining the reason for it there.

**Comment by ender-titan1** [comment]
I think mechanically, this feature is mostly ready. But since it still needs some work maybe it should be classed as an "experimental feature"?

To recap, what is needed is balance, AI implementation, a tutorial and maybe some graphics improvements.

**Comment by hhyyrylainen** [comment]
I suppose without those still undone parts classifying this as an experimental feature would be an option to merge this soon.

File: locale/fi.po [hunk 12]
```diff
@@ -5205,6 +5227,10 @@ msgstr ""
 msgid "TOGGLE_PAUSE"
 msgstr "Pysäytä"
 
+#, fuzzy
```
Comment (hhyyrylainen) [inline locale/fi.po:5230]: Could you redo the translations from scratch for this PR? I see this and the en.po file have a bunch of old translations added that will never be useful as the translation keys have never existed on the master branch.

File: src/microbe_stage/systems/StrainSystem.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
+﻿namespace Systems
+{
+    using Components;
+    using DefaultEcs;
+    using DefaultEcs.System;
+    using DefaultEcs.Threading;
+    using Godot;
+    using World = DefaultEcs.World;
+
+    [With(typeof(CompoundStorage))]
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:10]: This system seems to only use StrainAffected and OrganelleContainer components now, so the attributes here need to be fixed.

File: src/microbe_stage/components/StrainAffected.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    /// <summary>
+    ///   Contains variables related to strain
+    /// </summary>
+    [JSONDynamicTypeAllowed]
+    public struct StrainAffected
+    {
+        /// <summary>
+        ///   True when sprinting or when strain is supposed to be otherwise generated
+        /// </summary>
+        public bool IsUnderStrain;
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/StrainAffected.cs:14]: Could you reorder these variables for memory efficiency? i.e. biggest variables first and this bool last as this takes just one byte so this being first causes 3 extra padding bytes to be added (in this case this doesn't matter but if there was a second bool in this struct then this would consume extra memory unnecessarily so it is good practice).

File: src/microbe_stage/components/StrainAffected.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+        ///   The amount of time the player has to wait before <see cref="CurrentStrain"/> sarts to fall
+        /// </summary>
+        public float StrainDecreaseCooldown;
+    }
+
+    public static class StrainAffectedHelpers
+    {
+        public static float CalculateStrainFraction(this ref StrainAffected affected)
+        {
+            return Math.Max(0, affected.CurrentStrain - Constants.CANCELED_STRAIN) / Constants.MAX_STRAIN_PER_CELL;
```
Comment (hhyyrylainen) [inline src/microbe_stage/components/StrainAffected.cs:31]: Is the max strain per cell variable well named? I ask because for cell colonies they contain many cells. Should it perhaps be named `MAX_STRAIN_PER_ENTITY`?

File: src/microbe_stage/PlayerMicrobeInput.cs [hunk 16]
```diff
@@ -297,6 +297,40 @@ public void CloseSignalingCommandsMenu()
             stage.HUD.ApplySignalCommand(command, stage.Player);
     }
 
+    [RunOnKeyDown("g_sprint")]
+    public bool StartSprint()
+    {
+        if (!stage.HasPlayer || !stage.GameWorld.WorldSettings.ExperimentalFeatures)
+            return true;
```
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:304]: Shouldn't this *not* consume input, this doesn't do anything except cause other things to not be able to react to this keypress?

File: src/microbe_stage/PlayerMicrobeInput.cs [hunk 17]
```diff
@@ -297,6 +297,40 @@ public void CloseSignalingCommandsMenu()
...
+        control.Sprinting = true;
+
+        // We need to not consume the input, otherwise the key up for this will not run
+        return false;
+    }
+
+    [RunOnKeyUp("g_sprint")]
+    public void EndSprint()
+    {
+        if (!stage.HasPlayer || !stage.GameWorld.WorldSettings.ExperimentalFeatures)
```
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:316]: I don't think this needs to check experimental features, sprinting should always be able to end even if experimental features are not on.

File: src/microbe_stage/PlayerMicrobeInput.cs [hunk 18]
```diff
@@ -297,6 +297,40 @@ public void CloseSignalingCommandsMenu()
...
+            return;
+
+        ref var control = ref stage.Player.Get<MicrobeControl>();
+
+        control.Sprinting = false;
+    }
+
+    public void ToggleSprint()
+    {
+        if (!stage.HasPlayer)
```
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:326]: Instead checking here for experimental features would be more important I think.

File: src/general/base_stage/CreatureStageHUDBase.cs [hunk 19]
```diff
@@ -125,6 +125,9 @@ public abstract class CreatureStageHUDBase<TStage> : HUDWithPausing, ICreatureSt
     [Export]
     public NodePath HotBarPath = null!;
 
+    [Export]
+    public NodePath SprintHotkeyPath = null!;
```
Comment (hhyyrylainen) [inline src/general/base_stage/CreatureStageHUDBase.cs:129]: It seems the multicellular HUD scene is not setup for strain, it should be setup to ensure the game doesn't crash when getting to the prototypes.

File: src/general/base_stage/CreatureStageHUDBase.cs [hunk 20]
```diff
@@ -1052,14 +1057,17 @@ protected void UpdatePanelSizing(float delta)
     protected abstract void UpdateAbilitiesHotBar();
 
     protected void UpdateBaseAbilitiesBar(bool showEngulf, bool showToxin, bool showSlime,
-        bool showingSignaling, bool engulfOn, bool showEject)
+        bool showingSignaling, bool engulfOn, bool sprintOn, bool showEject)
     {
+        sprintHotkey.Visible = stage?.GameWorld.WorldSettings.ExperimentalFeatures == true;
```
Comment (hhyyrylainen) [inline src/general/base_stage/CreatureStageHUDBase.cs:1062]: Instead of allowing stage to be null here, I think a better thing to do would be to just flat out:
```
if (stage == null)
    throw new InvalidOperationException("Stage must be set before abilities update");
```

File: src/microbe_stage/MicrobeHUD.cs [hunk 21]
```diff
@@ -30,10 +30,22 @@ public class MicrobeHUD : CreatureStageHUDBase<MicrobeStage>
...
     public NodePath UnbindAllHotkeyPath = null!;
 
+    [Export]
+    public NodePath StrainBarPanelPath = null!;
+
+    [Export]
+    public NodePath StrainBarPath = null!;
+
+    [Export]
+    public NodePath StrainBarFadeAnimationPlayerPath = null!;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:40]: I think this should maybe be moved to the base HUD as shouldn't big creatures be eventually be able to sprint as well? So it makes sense to put there rather than having to copy-paste all this stuff to the multicellular HUD.

File: src/microbe_stage/MicrobeHUD.cs [hunk 22]
```diff
@@ -542,12 +566,58 @@ protected override void Dispose(bool disposing)
...
 
         base.Dispose(disposing);
     }
 
+    private void UpdateStrain(Entity player)
+    {
+        if (!stage!.GameWorld.WorldSettings.ExperimentalFeatures)
+        {
+            strainBar.Visible = false;
+        }
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:583]: Shouldn't there be an early return here?

File: src/microbe_stage/MicrobeHUD.cs [hunk 23]
```diff
@@ -542,12 +566,58 @@ protected override void Dispose(bool disposing)
...
+
+        var strainFraction = player.Get<StrainAffected>().CalculateStrainFraction();
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:585]: This should check if the component exists and if not print an error *once* and then do nothing. This will ensure compatibility with older saves and rare bugs if the component is somehow missing. This isn't a critical feature to display the info so it missing shouldn't crash the game.

File: project.godot [hunk 24]
```diff
@@ -476,6 +476,11 @@ g_eject_engulfed={
 "events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":false,"meta":false,"command":false,"pressed":false,"scancode":72,"physical_scancode":0,"unicode":0,"echo":false,"script":null)
  ]
 }
+g_sprint={
+"deadzone": 0.5,
+"events": [ Object(InputEventKey,"resource_local_to_scene":false,"resource_name":"","device":0,"alt":false,"shift":false,"control":false,"meta":false,"command":false,"pressed":false,"scancode":16777237,"physical_scancode":0,"unicode":0,"echo":false,"script":null)
```
Comment (hhyyrylainen) [inline project.godot:481]: What key is this? Would it be possible to make sprint be shift, like it is in most games?

File: src/microbe_stage/systems/StrainSystem.cs [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            ref var strain = ref entity.Get<StrainAffected>();
+            ref var organelles = ref entity.Get<OrganelleContainer>();
+
+            if (strain.IsUnderStrain)
+            {
+                var strainIncrease = Constants.SPRINTING_STRAIN_INCREASE_PER_SECOND * delta;
+                strainIncrease += organelles.HexCount * Constants.SPRINTING_STRAIN_INCREASE_PER_HEX;
+
+                if (strain.CurrentStrain > Constants.MAX_STRAIN_PER_CELL)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/StrainSystem.cs:32]: Shouldn't this check be below line 35 so that the new value is correctly clamped?

File: src/microbe_stage/systems/MicrobeMovementSystem.cs [hunk 26]
```diff
@@ -247,6 +273,12 @@ private static float CalculateRotationSpeed(in Entity entity, ref OrganelleConta
             return position.Rotation.Xform(movementVector);
         }
 
+        private float GetStrainMultiplier(ref StrainAffected strain)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeMovementSystem.cs:276]:
```suggestion
        private float GetStrainAtpMultiplier(ref StrainAffected strain)
```

File: src/microbe_stage/systems/MicrobeMovementSystem.cs [hunk 27]
```diff
@@ -140,8 +144,19 @@ private static float CalculateRotationSpeed(in Entity entity, ref OrganelleConta
...
+            ref var strain = ref entity.Get<StrainAffected>();
+
             if (control.MovementDirection == Vector3.Zero)
             {
+                // Make sure the microbe is not under strain when not moving
+                strain.IsUnderStrain = false;
+
+                // Remove ATP due to strain even if not moving
+                // This is calculated similarily to the regular movement cost for consistency
+                var strainMultiplier = GetStrainMultiplier(ref strain);
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/MicrobeMovementSystem.cs:156]: Couldn't this line be straight below the `ref var strain` line as this value seems to be always needed? That way this doesn't need to have the code to calculate this in two places.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Here's a full code review. I don't see any major issues anymore (unless some pop up when I playtest this), but there's quite many small details that can be tweaked to make this feature more maintainable and improve code clarity, and maybe fix one small bug.

File: locale/fi.po (hunk 12 above)
Comment (hhyyrylainen) [reply locale/fi.po:5230]: Also I see this has translation file conflicts so anyway redoing the translations shouldn't be harder than fixing those conflicts manually.

File: project.godot (hunk 24 above)
Comment (ender-titan1) [reply project.godot:481]: This is shift. From my testing it seems to work.

Comment (hhyyrylainen) [reply project.godot:481]: Alright, that's great then :+1:

File: src/microbe_stage/MicrobeHUD.cs [hunk 28]
```diff
@@ -311,6 +317,18 @@ protected override void UpdateHealth(float delta)
         hpLabel.HintTooltip = hpText;
     }
 
+    protected override float? ReadPlayerStrainFraction()
+    {
+        if (stage!.Player.Has<StrainAffected>() && !playerMissingStrainAffected)
```
Comment (CI09) [inline src/microbe_stage/MicrobeHUD.cs:322]: I think first condition should be with ! at the beggining

File: src/microbe_stage/components/MicrobeControl.cs [hunk 29]
```diff
@@ -53,6 +53,11 @@ public struct MicrobeControl
         /// </summary>
         public bool SlowedBySlime;
 
+        /// <summary>
+        ///   Whether this microbe is currenyly sprinting
```
Comment (CI09) [inline src/microbe_stage/components/MicrobeControl.cs:57]: Typo

**Review by CI09 — CHANGES_REQUESTED** [review]

**Comment by Deus-Codes** [comment]
Limited testing due to battery issues with my laptop. I am not sure how it scales with greater mass, but sprinting felt pretty good in the few generations I played. Great job implementing a feature that makes movement atleast a bit more engaging than pointing your mouse in a direction and pressing W.

One thing though: in the original concept, sprinting was attached to the flagella as an ability rather than a default ability, which synergizes with the flagella providing upgrades changing the behavior of sprinting. Was it an intentional decision to have it be available from the get go?

**Comment by ender-titan1** [comment]
Oh I wasn't expecting this to be reviewed. Slight heads up that this will probably have to be changed into a new Godot 4 branch. I don't recall there being a particular reason for why the sprinting is enabled by default, iirc it was just an arbitrary decision that's pretty easy to change if needed. 

However I **did** intentionally leave out flagella upgrades for the simple reason that I found them too hard to implement and decided to do them down the line when the basic system was done.

**Comment by hhyyrylainen** [comment]
Will this be continued soon? It's not going to be too much longer now that the 0.6.x roadmap items are all planned to be completed and this is one of them.
